pkg/sql/regionliveness/prober.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package regionliveness
+
+import (
+	"context"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	clustersettings "github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slinstance"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+)
+
+var RegionLivenessEnabled = settings.RegisterBoolSetting(settings.ApplicationLevel,
+	"sql.region_liveness.enabled",
+	"enables region liveness for system databases (experimental)",
+	false, /* disabled */
+	settings.WithVisibility(settings.Reserved))
+
+// LiveRegions are regions which are currently still avaialble,
+// and not quarantined due to expiration.
+type LiveRegions map[string]struct{}
+
+// Prober used to determine the set of regions which are still alive.
+type Prober interface {
+	// ProbeLiveness can be used after a timeout to label a regions as unavailable.
+	ProbeLiveness(ctx context.Context, region string) error
+	// QueryLiveness can be used to get the list of regions which are currently
+	// accessible.
+	QueryLiveness(ctx context.Context, txn *kv.Txn) (LiveRegions, error)
+}
+
+type livenessProber struct {
+	db       isql.DB
+	settings *clustersettings.Settings
+}
+
+// NewLivenessProber creates a new region liveness prober.
+func NewLivenessProber(db isql.DB, settings *clustersettings.Settings) Prober {
+	return &livenessProber{
+		db:       db,
+		settings: settings,
+	}
+}
+
+// fetchSystemDBRegions fetches the regions that are added in the system database.
+func (l *livenessProber) fetchSystemDBRegions(
+	ctx context.Context, executor isql.Executor, txn *kv.Txn,
+) ([]string, error) {
+	const fetchDbRegions = `
+SELECT
+    regions 
+FROM system.crdb_internal.databases
```

Comment:
Instead of using the `crdb_internal` virtual table, I think this should use `regions/region_provider.go`. It's a bit more efficient and makes it easier to reason about the behavior of fetchSystemDBRegions while the sql server is bootstrapping.